fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePrimaryKeySegmentIndex.java

```diff
@@ -0,0 +1,335 @@
+                final long mergedSegmentId = directory.primaryKeySegmentId(mergedSegmentInfo.name, false);
+                final Bits liveDocs = mergeState.liveDocs[i];
+                final MergeState.DocMap docMap = mergeState.docMaps[i];
+                final int maxDoc = mergeState.maxDocs[i];
+                for (int j = 0; j < maxDoc; j++) {
+                    storedFieldsReader.visitDocument(j, visitor);
+                    final byte[] primaryKey = visitor.getPrimaryKey();
+                    if (primaryKey != null) {
+                        if (liveDocs == null || liveDocs.get(j)) {
+                            int docId = docMap.get(j);
+                            if (docId >= 0) {
+                                addOrDeletePrimaryKeyEntry(primaryKey, segmentId, docId, true);
+                            }
+                        }
+                        // TODO: Is doing this right away safe enough? Does transaction isolation protect against all cases where
+                        //  the source segment is used after a merge?
+                        addOrDeletePrimaryKeyEntry(primaryKey, mergedSegmentId, j, false);
+                        visitor.reset();
+                    }
+                }
+            }
+
+            directory.getContext().increment(LuceneEvents.Counts.LUCENE_MERGE_DOCUMENTS, docCount);
+            directory.getContext().increment(LuceneEvents.Counts.LUCENE_MERGE_SEGMENTS, segmentCount);
+
+            return docCount;
+        }
+
+        @Override
+        public void finish(FieldInfos fis, int numDocs) throws IOException {
+            inner.finish(fis, numDocs);
+        }
+
+        @Override
+        public void close() throws IOException {
+            inner.close();
+        }
+
+        @Override
+        public long ramBytesUsed() {
+            return inner.ramBytesUsed();
+        }
+
+        @Override
+        public Collection<Accountable> getChildResources() {
+            return inner.getChildResources();
+        }
+    }
+
+    /**
+     * Reconcile primary key index entries for a segment.
+     * Delete those for a segment that is about to be deleted or add for a segment that has been filled without going through the wrapper.
+     * The primary keys are gotten from the stored fields, so this must complete and not yet have been removed.
+     * @param reader a reader for the target segment
+     * @param add {@code true} to add entries; {@code false} to delete them
+     * @throws IOException if the stored fields gets an error
+     */
+    @SuppressWarnings("PMD.CloseResource")
+    public void addOrDeleteDocumentPrimaryKeys(@Nonnull SegmentReader reader, boolean add) throws IOException {
+        final int maxDoc = reader.maxDoc();
+        final PrimaryKeyVisitor visitor = new PrimaryKeyVisitor();
+        final long segmentId = directory.primaryKeySegmentId(reader.getSegmentName(), add);
```

Comment:
Instead of doing this can you get the segmentId from `reader.getSegmentInfo`?

Reply:
I thought Lucene had a number associated with segments, but I can't find it now. It has an `id`, but that's based off a random `BigInteger`, so not really helpful.